### PR TITLE
Add audit logging to CRUD actions using Django signals

### LINKED ISSUE #3178: Log trails on CRUD actions within Rocky models
### Detailed description

In order to have more audit trails, we should introduce logging based on CRUD actions of the models. This could be achieved by hooking onto specific signals of specific models. We should also explore if and how we can add request and user information to CRUD actions. This would provide insights into the context and the user responsible for creating or updating a specific model.

This feature would be a continuation of #3114 

### Specifications

The implementation should include:

- [ ] Logging on CRUD operations of specific models

## Implementation

- Hook onto signals and log actions

### PR DESCRIPTION
**Pull request by dekkers**
### Changes

This adds audit logging for CRUD actions in Django using signals.

The check for token auth so that tokens can be used for non-DRF views has been moved to its own middleware. The auth required middleware should come after the structlog middleware, but the new auth token middleware is put before the structlog middleware so that middleware can use the user.

### Issue link

Closes #3178

### Demo

image

### QA notes

Check that adding, changing and removing rocky objects (users, organizations, members, indemnifications, etc.) still work and log the right message.

---

## Checklist for code reviewers:

Copy-paste the checklist from [the docs/source/templates folder](https://github.com/minvws/nl-kat-coordination/blob/main/docs/source/templates/pull_request_template_review_code.md) into your comment.

---

## Checklist for QA:

Copy-paste the checklist from [the docs/source/templates folder](https://github.com/minvws/nl-kat-coordination/blob/main/docs/source/templates/pull_request_template_review_qa.md) into your comment.

### CODE REVIEW

File: rocky/rocky/signals.py [hunk 1]
```diff
@@ -1,5 +1,38 @@
...
-task_received = Signal()
-task_succeeded = Signal()
-task_failed = Signal()
+logger = structlog.get_logger(__name__)
+
+
+def log_save(sender, instance, created, **kwargs) -> None:
+    if isinstance(instance, LogEntry):
+        # Django admin will automatically create a LogEntry for each admin
+        # action, but we shouldn't send log messages about thosee.
```
Comment (Donnype) [inline rocky/rocky/signals.py:10]:
```suggestion
        # action, but we shouldn't send log messages about these.
```

**Comment by Donnype** [comment]
> Can we add the EventId's as mentioned in the issue?

To save others some clicks:
- https://github.com/minvws/nl-kat-coordination/issues/3043

Do you mean "Event Code" in this case @underdarknl?

File: rocky/rocky/apps.py [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+class RockyConfig(AppConfig):
+    name = "rocky"
+
+    def ready(self) -> None:
+        # We need to do the import here because else we will get an "apps aren't
+        # loaded yet" error from Django.
+        from rocky.signals import log_delete, log_save
+
+        post_save.connect(log_save, dispatch_uid="log_save")
+        post_delete.connect(log_delete, dispatch_uid="log_delete")
```
Comment (ammar92) [inline rocky/rocky/apps.py:14]: What's the purpose of the `dispatch_uid`?

File: rocky/rocky/middleware/auth_token.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    def middleware(request):
+        if not request.user.is_authenticated and "authorization" in request.headers:
+            authenticator = TokenAuthentication()
+            try:
+                user, token = authenticator.authenticate(request)
+            except APIException:
+                return HttpResponseForbidden("Invalid token\n")
+            else:
+                request.user = user
+                structlog.contextvars.bind_contextvars(auth_method="token")
```
Comment (ammar92) [inline rocky/rocky/middleware/auth_token.py:17]: Nice one

**Review by ammar92 — APPROVED** [review]

**Comment by stephanie0x00** [comment]
## Checklist for QA:

- [x] I have checked out this branch, and successfully ran a fresh `make reset`.
- [x] I confirmed that there are no unintended functional regressions in this branch:
  - [x] I have managed to pass the onboarding flow
  - [x] Objects and Findings are created properly
  - [x] Tasks are created and completed properly
- [ ] I confirmed that the PR's advertised `feature` or `hotfix` works as intended.
- [x] I checked the logs for errors and/or warnings and made issues where necessary

### What works:
When creating organisations and adding users the log message. 

Organisation created: 
```
rocky-1  | {"event": "{\"object_type\": \"Organization\", \"object\": \"aa\", \"event\": \"Organization aa created\", \"ip\": \"172.30.0.1\", \"user_id\": 1, \"request_id\": \"8b6c58bd-9d6c-4400-9a2b-5cb64ef055cc\", \"level\": \"info\", \"timestamp\": \"2024-08-05T09:20:43.828375\"}"}
```

Member created:
```
rocky-1  | {"event": "{\"object_type\": \"OrganizationMember\", \"object\": \"superuser@localhost\", \"event\": \"OrganizationMember superuser@localhost created\", \"ip\": \"172.30.0.1\", \"user_id\": 1, \"request_id\": \"8b6c58bd-9d6c-4400-9a2b-5cb64ef055cc\", \"level\": \"info\", \"timestamp\": \"2024-08-05T09:20:43.840083\"}"}
```

Member updated:
```
rocky-1  | {"event": "{\"object_type\": \"OrganizationMember\", \"object\": \"superuser@localhost\", \"event\": \"OrganizationMember superuser@localhost updated\", \"ip\": \"172.30.0.1\", \"user_id\": 1, \"request_id\": \"8b6c58bd-9d6c-4400-9a2b-5cb64ef055cc\", \"level\": \"info\", \"timestamp\": \"2024-08-05T09:20:43.841961\"}"}
```

Indemnification event: 
```
rocky-1  | {"event": "{\"object_type\": \"Indemnification\", \"object\": \"Indemnification object (2)\", \"event\": \"Indemnification Indemnification object (2) created\", \"ip\": \"172.30.0.1\", \"user_id\": 1, \"request_id\": \"43d8d964-1402-47b1-8a97-315301776805\", \"level\": \"info\", \"timestamp\": \"2024-08-05T09:25:39.773688\"}"}
```

### What doesn't work:
See below. 

### Bug or feature?:
- When updating a user account, eg. make the user account blocked, the logs only show that the specific user got updated, without context as to what was updated. Is this expected/intended behaviour? 
- When deleting an ooi should this also trigger a deletion event? 
- Indemnification evens miss some context, like which organization is this applicable to? Who triggered this? Was this an automatic change by the system or a manual indemnification change?

**Comment by dekkers** [comment]
> ### What doesn't work:
> 
> See below.
> ### Bug or feature?:
> 
>     * When updating a user account, eg. make the user account blocked, the logs only show that the specific user got updated, without context as to what was updated. Is this expected/intended behaviour?

This is expected behaviour. The signal only gets the information that the account object was updated. If we want to log more information, we need to do that in the view.

>     * When deleting an ooi should this also trigger a deletion event?

Yes, but that is not a Django model CRUD action, so logging this can not be done in this signal handler and was not part of this ticket.

>     * Indemnification evens miss some context, like which organization is this applicable to? Who triggered this? Was this an automatic change by the system or a manual indemnification change?

The signal handler logs the standard string representation of the Django model. For the indemnifcation the standard string representation doesn't contain the organisation which we should fix. Other models might also not a have a good string representation and I think we should fix that in a new ticket.

**Comment by underdarknl** [comment]
> The signal handler logs the standard string representation of the Django model. For the indemnifcation the standard string representation doesn't contain the organisation which we should fix. Other models might also not a have a good string representation and I think we should fix that in a new ticket.

Without the organisation, the log is meaningless.

**Pull request merged**